### ISSUE TITLE
Repair t_dgram_worker test

### ISSUE THREAD

**Issue by adminspotter** [adminspotter/r9#168]
The test succeeds sometimes and fails other times on our CI platform.

```
Running main() from gtest_main.cc
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from DgramSocketTest
[ RUN      ] DgramSocketTest.ListenWorker
created dgram socket 3 on port 8765
0
1
2
3
invalid address family 0
4
exiting connection loop for datagram port 8765
[       OK ] DgramSocketTest.ListenWorker (0 ms)
[ RUN      ] DgramSocketTest.SendWorker
created dgram socket 3 on port 8764
starting connection loop for datagram port 8764
started reaper thread for started access pool worker for datagram port 8764
started send pool worker for datagram port 8764
error sending packet out datagram port 8764: Invalid argument (22)
exiting connection loop for datagram port 8764
pure virtual method called
terminate called without an active exception
```

It's not clear which pure-virtual method we're calling, but the error is arising when calling the destructor.

**Comment by adminspotter** [comment]
The destructor calls stop(), which may call port_type(), a pure-virtual.  As the virtual destructors unwind the object, there is a time in which a dgram or stream socket isn't really either of those things, and will be called from _only_ a listen_socket context, thus the call out to the pure-virtual.

Moving the stop() calls into the derived classes' destructors doesn't seem to help.  If we want that variable behaviour depending on what subclass is calling, we'll have to implement that for real in the base class.

There are probably other ways around it that can maintain a good level of code reuse, but they're not being obvious at the moment.

**Comment by adminspotter** [comment]
I'm considering this fully resolved in 4571b14c6161dd6edfec68e0b138b7cc13ba33ef, but won't be merged into master until we finish with #172.
